src/cachingreader.cpp

```diff
@@ -162,66 +159,90 @@ void CachingReader::freeAllChunks() {
+            pChunk->state = Chunk::FREE;
+            pChunk->chunk_number = -1;
+            pChunk->length = 0;
+            pChunk->next_lru = NULL;
+            pChunk->prev_lru = NULL;
+            m_freeChunks.append(pChunk);
         }
     }
 }
 
-Chunk* CachingReader::allocateChunk() {
-    if (m_freeChunks.count() == 0)
+Chunk* CachingReader::allocateChunk(int chunk) {
+    if (m_freeChunks.isEmpty()) {
         return NULL;
-    return m_freeChunks.takeFirst();
+    }
+    Chunk* pChunk = m_freeChunks.takeFirst();
+    pChunk->state = Chunk::ALLOCATED;
+    pChunk->chunk_number = chunk;
+
+    //qDebug() << "Inserting chunk" << pChunk << pChunk->chunk_number;
+    m_allocatedChunks.insert(pChunk->chunk_number, pChunk);
+
+    // Insert the chunk into the LRU list
+    m_mruChunk = insertIntoLRUList(pChunk, m_mruChunk);
+
+    // If this chunk has no next LRU then it is the LRU. This only
+    // happens if this is the first allocated chunk.
+    if (pChunk->next_lru == NULL) {
+        m_lruChunk = pChunk;
+    }
+    return pChunk;
 }
 
-Chunk* CachingReader::allocateChunkExpireLRU() {
-    Chunk* chunk = allocateChunk();
-    if (chunk == NULL) {
+Chunk* CachingReader::allocateChunkExpireLRU(int chunk) {
+    Chunk* pChunk = allocateChunk(chunk);
+    if (pChunk == NULL) {
         if (m_lruChunk == NULL) {
             qDebug() << "ERROR: No LRU chunk to free in allocateChunkExpireLRU.";
             return NULL;
         }
         //qDebug() << "Expiring LRU" << m_lruChunk << m_lruChunk->chunk_number;
         freeChunk(m_lruChunk);
-        chunk = allocateChunk();
+        pChunk = allocateChunk(chunk);
     }
-    //qDebug() << "allocateChunkExpireLRU" << chunk;
-    return chunk;
+    //qDebug() << "allocateChunkExpireLRU" << chunk << pChunk;
+    return pChunk;
 }
 
 Chunk* CachingReader::lookupChunk(int chunk_number) {
     // Defaults to NULL if it's not in the hash.
-    Chunk* chunk = NULL;
+    Chunk* chunk = m_allocatedChunks.value(chunk_number, NULL);
 
-    if (m_allocatedChunks.contains(chunk_number)) {
-        chunk = m_allocatedChunks.value(chunk_number);
+    // Make sure the allocated number matches the indexed chunk number.
+    DEBUG_ASSERT(chunk == NULL || chunk_number == chunk->chunk_number);
 
-        // Make sure we're all in agreement here.
-        if (chunk_number != chunk->chunk_number) {
-            qDebug() << "ERROR: Inconsistent chunk has chunk_number that doesn't match allocated-chunks key.";
-        }
+    return chunk;
+}
 
-        // If this is the LRU chunk then set the previous LRU to the new LRU
-        if (chunk == m_lruChunk && chunk->prev_lru != NULL) {
-            m_lruChunk = chunk->prev_lru;
-        }
-        // Remove the chunk from the list and insert it at the head.
-        m_mruChunk = removeFromLRUList(chunk, m_mruChunk);
-        m_mruChunk = insertIntoLRUList(chunk, m_mruChunk);
+void CachingReader::freshenChunk(Chunk* pChunk) {
```

Comment:
It takes the chunk and re-inserts it in the least-recently-used list so it doesn't read the chunk. I chose freshen as in "freshen the chunk in the cache" since stale / least-recently-used items are expired. "recache" was another options I considered but that sounded odd to me. Any ideas?